chain/client/src/chunk_inclusion_tracker.rs

```diff
@@ -1,37 +1,60 @@
+use chrono::{DateTime, Utc};
 use itertools::Itertools;
 use lru::LruCache;
+use near_primitives::checked_feature;
+use near_primitives::version::ProtocolVersion;
 use std::collections::HashMap;
+use std::sync::Arc;
 
+use near_chain_primitives::Error;
+use near_epoch_manager::EpochManagerAdapter;
+use near_primitives::block_body::ChunkEndorsementSignatures;
 use near_primitives::hash::CryptoHash;
-use near_primitives::sharding::ShardChunkHeader;
+use near_primitives::sharding::{ChunkHash, ShardChunkHeader};
 use near_primitives::types::{AccountId, EpochId, ShardId};
 
+use crate::chunk_validation::ChunkValidator;
 use crate::metrics;
 
 const CHUNK_HEADERS_FOR_INCLUSION_CACHE_SIZE: usize = 2048;
 const NUM_EPOCH_CHUNK_PRODUCERS_TO_KEEP_IN_BLOCKLIST: usize = 1000;
 
+struct ChunkInfo {
+    // chunk_header, received_time and chunk_producer are populated when we call mark_chunk_header_ready_for_inclusion
+    pub chunk_header: ShardChunkHeader,
+    pub received_time: DateTime<Utc>,
+    pub chunk_producer: AccountId,
+    // The signatures are populated later during call to chunk_validator.get_chunk_endorsement_signature
+    // This is a sort of cache to store previously fetched signatures.
+    pub signatures: Option<ChunkEndorsementSignatures>,
+}
+
 pub struct ChunkInclusionTracker {
     // Track chunks that are ready to be included in a block.
     // Key is the previous_block_hash as the chunk is created based on this block. It's possible that
     // the block included isn't of height previous_block_height + 1 in cases of skipped blocks etc.
-    // We store the map of chunks from [shard_id] to (chunk_header, chunk received_time, chunk_producer account_id)
-    prev_block_to_chunk_headers_ready_for_inclusion: LruCache<
-        CryptoHash,
-        HashMap<ShardId, (ShardChunkHeader, chrono::DateTime<chrono::Utc>, AccountId)>,
-    >,
+    // We store the map of chunks from [shard_id] to chunk_hash
+    prev_block_to_chunk_hash_ready: LruCache<CryptoHash, HashMap<ShardId, ChunkHash>>,
+
+    // Map from chunk_hash to chunk_info.
+    // ChunkInfo stores the chunk_header, received_time, chunk_producer and chunk endorsements.
+    // Cleaning up of chunk_hash_to_chunk_info is handled during cache eviction from prev_block_to_chunk_hash_ready.
+    chunk_hash_to_chunk_info: HashMap<ChunkHash, ChunkInfo>,
 
     // Track banned chunk producers for a given epoch. We filter out chunks produced by them.
     banned_chunk_producers: LruCache<(EpochId, AccountId), ()>,
+
+    // Epoch manager to get the protocol version for a given epoch.
+    epoch_manager: Arc<dyn EpochManagerAdapter>,
```

Comment:
random better engineering idea, not for this pr
This seems quite heavy, to pass the whole epoch manager just to get the protocol version for epoch. I'm guessing it makes testing a bit more involved too. If this is a recurring pattern we can subdivide epoch manager into smaller traits.

Reply:
Testing is relatively ok actually, Most of EpochManager is just a ton of logic that sits on top of a datastore of BlockInfo. The setup of EpochManager isn't too bad as far as I remember, and splitting that might actually be counterproductive because it can make tests invent setups that aren't possible with a real set of validators.